File: box/src/main/java/de/qabel/box/storage/BoxExternalReference.java
package de.qabel.box.storage;

import de.qabel.core.crypto.QblECPublicKey;

import java.util.Arrays;

public class BoxExternalReference {
    public boolean isFolder;
    public String name;
    public String url;
    public QblECPublicKey owner;
    public byte[] key;

    public BoxExternalReference(boolean isFolder, String url, String name, QblECPublicKey owner, byte[] key) {
        this.isFolder = isFolder;
        this.name = name;
        this.url = url;
        this.owner = owner;
        this.key = key;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        BoxExternalReference that = (BoxExternalReference) o;

        if (isFolder != that.isFolder) {
            return false;
        }
        if (url != null ? !url.equals(that.url) : that.url != null) {
            return false;
        }
        if (owner != null ? !owner.equals(that.owner) : that.owner != null) {
            return false;
        }
        return Arrays.equals(key, that.key);

    }

    @Override
    public int hashCode() {
        int result = isFolder ? 1 : 0;
        result = 31 * result + (url != null ? url.hashCode() : 0);
        result = 31 * result + (owner != null ? owner.hashCode() : 0);
        result = 31 * result + (key != null ? Arrays.hashCode(key) : 0);
        return result;
    }

    @Override
    protected BoxExternalReference clone() throws CloneNotSupportedException {
        return new BoxExternalReference(isFolder, url, name, owner, key);
    }
}

File: box/src/main/java/de/qabel/box/storage/BoxExternalReference.kt
package de.qabel.box.storage

import de.qabel.core.crypto.QblECPublicKey

import java.util.Arrays

class BoxExternalReference(var isFolder: Boolean, var url: String?, var name: String, var owner: QblECPublicKey?, var key: ByteArray?) {

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }

        val that = o as BoxExternalReference?

        if (isFolder != that!!.isFolder) {
            return false
        }
        if (if (url != null) url != that.url else that.url != null) {
            return false
        }
        if (if (owner != null) owner != that.owner else that.owner != null) {
            return false
        }
        return Arrays.equals(key, that.key)

    }

    override fun hashCode(): Int {
        var result = if (isFolder) 1 else 0
        result = 31 * result + if (url != null) url!!.hashCode() else 0
        result = 31 * result + if (owner != null) owner!!.hashCode() else 0
        result = 31 * result + if (key != null) Arrays.hashCode(key) else 0
        return result
    }

    @Throws(CloneNotSupportedException::class)
    protected fun clone(): BoxExternalReference {
        return BoxExternalReference(isFolder, url, name, owner, key)
    }
}
